### ISSUE TITLE
Errors in PrepSCTFindMarkers and FindSpatiallyVariableFeatures

### ISSUE THREAD

**Issue by tingchiafelix** [satijalab/seurat#8235]
Hi,

First of all, thank you so much for providing this very useful tool. I've been using Seurat package since version 3, and recently jumped into Visium data so I'm still learning how to process the Visium spatial data.

The "Working with multiple slices in Seurat" section of the spatial vignette webpage elaborates very well on how to merge the multiple slices in Seurat and generate the feature on Spatial Dim plots. I was attempting to find spatially variable features across clusters.

There are two ways that I'm aware of using "1. FindAllMarkers" and "2. FindSpatiallyVariableFeatures" (#7094 ). I've been reading through the previous issues and questions, but unfortunately couldn't find the proper solutions.

Any suggestion would be very appreciated! 

Best,
TC

Below is my script:

```
a <- SCTransform(a, assay = "Spatial", verbose = FALSE)
b <- SCTransform(b, assay = "Spatial", verbose = FALSE)
c <- SCTransform(c, assay = "Spatial", verbose = FALSE)

HN_visium.merge <- merge(x = a, 
                  y = c(b,c))

DefaultAssay(HN_visium.merge) <- "SCT"
VariableFeatures(HN_visium.merge) <- c(VariableFeatures(a),VariableFeatures(b),VariableFeatures(c))

HN_visium.merge <- SCTransform(HN_visium.merge, assay = "Spatial", verbose = TRUE)
HN_visium.merge <- RunPCA(HN_visium.merge, verbose = FALSE)
HN_visium.merge <- FindNeighbors(HN_visium.merge, dims = 1:30)
HN_visium.merge <- FindClusters(HN_visium.merge, verbose = FALSE)
HN_visium.merge <- RunUMAP(HN_visium.merge, dims = 1:30)
```

```
1.
HN_visium.merge<- PrepSCTFindMarkers(HN_visium.merge, assay = "SCT", verbose = TRUE)

=> Before launching "FindAllMarkers", it requires running "PrepSCTFindMarkers" and I got the error below.

**Found 3 SCT models. Recorrecting SCT counts using minimum median counts: 6423.5
Error in PrepSCTFindMarkers(HN_visium.merge, assay = "SCT", verbose = TRUE) : 
  Multiple UMI assays are used for SCTransform: Spatial, RNA**

HN_visium.markers <- FindAllMarkers(HN_visium.merge, only.pos = TRUE)

```

```
2.
Initially, I got the error below and took the suggestion from issue #4611 but I had another error shown below by running "FindSpatiallyVariableFeatures".

HN_visium.merge <- FindSpatiallyVariableFeatures(HN_visium.merge, assay = "SCT", 
                                                 features = VariableFeatures(HN_visium.merge)[1:1000],
                                                 selection.method = "moransi")

1st error:
**Error in (function (cond)  : 
  error in evaluating the argument 'x' in selecting a method for function 'as.matrix': subscript out of bounds**

2nd error: (this error has been discussed on many posts but still can't figured out.)
**Error in FindSpatiallyVariableFeatures.default(object = data, spatial.location = spatial.location,  : 
  Please provide the same number of observations as spatial locations.**
```

```
R version 4.2.2 (2022-10-31)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS 14.2.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] Seurat_5.0.1                SeuratObject_5.0.1          sp_2.1-2                    SpatialExperiment_1.8.1    
 [5] SingleCellExperiment_1.20.1 SummarizedExperiment_1.28.0 Biobase_2.58.0              GenomicRanges_1.50.2       
 [9] GenomeInfoDb_1.34.9         IRanges_2.32.0              S4Vectors_0.36.2            BiocGenerics_0.44.0        
[13] MatrixGenerics_1.10.0       matrixStats_1.1.0           STdeconvolve_1.3.1          dplyr_1.1.4                
[17] patchwork_1.1.3             ggplot2_3.4.4              

loaded via a namespace (and not attached):
  [1] spam_2.10-0               plyr_1.8.9                igraph_1.6.0              lazyeval_0.2.2           
  [5] splines_4.2.2             RcppHNSW_0.5.0            BiocParallel_1.32.6       listenv_0.9.0            
  [9] scattermore_1.2           digest_0.6.33             htmltools_0.5.7           magick_2.8.2             
 [13] fansi_1.0.6               magrittr_2.0.3            tensor_1.5                cluster_2.1.6            
 [17] ROCR_1.0-11               limma_3.54.2              globals_0.16.2            R.utils_2.12.3           
 [21] spatstat.sparse_3.0-3     colorspace_2.1-0          ggrepel_0.9.4             crayon_1.5.2             
 [25] RCurl_1.98-1.13           jsonlite_1.8.8            progressr_0.14.0          spatstat.data_3.0-3      
 [29] survival_3.5-7            zoo_1.8-12                glue_1.6.2                polyclip_1.10-6          
 [33] gtable_0.3.4              zlibbioc_1.44.0           XVector_0.38.0            leiden_0.4.3.1           
 [37] DelayedArray_0.24.0       DropletUtils_1.18.1       Rhdf5lib_1.20.0           future.apply_1.11.0      
 [41] HDF5Array_1.26.0          abind_1.4-5               scales_1.3.0              edgeR_3.40.2             
 [45] spatstat.random_3.2-2     miniUI_0.1.1.1            Rcpp_1.0.11               viridisLite_0.4.2        
 [49] xtable_1.8-4              reticulate_1.34.0         dqrng_0.3.2               bit_4.0.5                
 [53] dotCall64_1.1-1           htmlwidgets_1.6.4         MERINGUE_1.0              httr_1.4.7               
 [57] RColorBrewer_1.1-3        ellipsis_0.3.2            ica_1.0-3                 farver_2.1.1             
 [61] pkgconfig_2.0.3           R.methodsS3_1.8.2         scuttle_1.8.4             uwot_0.1.16              
 [65] deldir_2.0-2              locfit_1.5-9.8            utf8_1.2.4                labeling_0.4.3           
 [69] tidyselect_1.2.0          rlang_1.1.2               reshape2_1.4.4            later_1.3.2              
 [73] munsell_0.5.0             tools_4.2.2               cli_3.6.2                 generics_0.1.3           
 [77] ggridges_0.5.5            stringr_1.5.1             fastmap_1.1.1             goftest_1.2-3            
 [81] bit64_4.0.5               fitdistrplus_1.1-11       purrr_1.0.2               RANN_2.6.1               
 [85] pbapply_1.7-2             future_1.33.0             nlme_3.1-164              sparseMatrixStats_1.10.0 
 [89] mime_0.12                 R.oo_1.25.0               hdf5r_1.3.8               compiler_4.2.2           
 [93] rstudioapi_0.15.0         plotly_4.10.3             png_0.1-8                 spatstat.utils_3.0-4     
 [97] tibble_3.2.1              glmGamPoi_1.10.2          stringi_1.8.3             RSpectra_0.16-1          
[101] lattice_0.22-5            Matrix_1.6-4              vctrs_0.6.5               pillar_1.9.0             
[105] lifecycle_1.0.4           rhdf5filters_1.10.1       spatstat.geom_3.2-7       lmtest_0.9-40            
[109] RcppAnnoy_0.0.21          data.table_1.14.10        cowplot_1.1.2             bitops_1.0-7             
[113] irlba_2.3.5.1             httpuv_1.6.13             R6_2.5.1                  promises_1.2.1           
[117] KernSmooth_2.23-22        gridExtra_2.3             parallelly_1.36.0         codetools_0.2-19         
[121] fastDummies_1.7.3         MASS_7.3-60               rhdf5_2.42.1              rjson_0.2.21             
[125] withr_2.5.2               presto_1.0.0              sctransform_0.4.1         GenomeInfoDbData_1.2.9   
[129] mgcv_1.9-1                parallel_4.2.2            grid_4.2.2                beachmat_2.14.2          
[133] tidyr_1.3.0               DelayedMatrixStats_1.20.0 Rtsne_0.17                spatstat.explore_3.2-5   
[137] shiny_1.8.0              
```

**Comment by saketkc** [comment]
Hi @tingchiafelix, 

My recommendation would be to reanme your "Spatial" assay to "RNA" assay and retry the steps.

**Comment by isrgbf** [comment]
> Hi @tingchiafelix,
> 
> My recommendation would be to reanme your "Spatial" assay to "RNA" assay and retry the steps.

I am currently stuck with the same issue, and can't seem to rename the assay. My error code is:
Found 2 SCT models. Recorrecting SCT counts using minimum median counts: 8472
Error in PrepSCTFindMarkers(dev0_merged_harmony, assay = "SCT", verbose = TRUE) : 
  Multiple UMI assays are used for SCTransform: Spatial, RNA

And I tried using RenameAssay() to make that switch, but didn't seem to work. I also tried using the argument for PrepSCTFindMarkers using object[["Spatial"]] and object[["RNA"]] but that also did not work.
What would be the correct way to approach this?

**Comment by StephanieHowe** [comment]
> > Hi @tingchiafelix,
> > My recommendation would be to reanme your "Spatial" assay to "RNA" assay and retry the steps.
> 
> I am currently stuck with the same issue, and can't seem to rename the assay. My error code is: Found 2 SCT models. Recorrecting SCT counts using minimum median counts: 8472 Error in PrepSCTFindMarkers(dev0_merged_harmony, assay = "SCT", verbose = TRUE) : Multiple UMI assays are used for SCTransform: Spatial, RNA
> 
> And I tried using RenameAssay() to make that switch, but didn't seem to work. I also tried using the argument for PrepSCTFindMarkers using object[["Spatial"]] and object[["RNA"]] but that also did not work. What would be the correct way to approach this?



Hi there,
I've also been struggling with the same issue for a few days and think I finally figured it out. For some reason, it seems that SCTransform changes the "umi.assay" slot of all the SCTAssay objects in a merged object (except the first one) to "RNA". I changed this for each component of my merged object using the following, and it didn't seem to affect anything downstream. This is what I used (my merged object is creatively called merge): 
slot(object = merge@assays$SCT@SCTModel.list[[2]], name="umi.assay")<-"Spatial"

You can then check this with:
SCTResults(object=merge, slot="umi.assay")

Hope this helps

**Comment by groverj3** [comment]
I know this issue is closed, but should it be?

I merged multiple Xenium samples and all but the first had their assays renamed to "RNA." This issue was my clue that I'd need to make them consistent in order for downstream functions to work properly. I ran into the same error in `PrepSCTFindMarkers()` as above.

I would think this is a bug and multi-sample spatial experiments won't work properly without a manual intervention to rename the assays as shown here.
